### PR TITLE
OCPBUGS-15331,OCPBUGS-16049: Enable AdvertiseAddress dual stack and IPv6 support and added the changes to be included in the certificates

### PR DESCRIPTION
**Pull request by jparrill**
## Changes:

- Now IPv6 and Dual Stack scenarios are covered in self-hosted environments.
- KAS PKI modified to include IPv4/6 addresses into the certificate
- Proxy still support only single stack scenarios, using the first ServiceCIDR as a AdvertiseAddress or InternalAPIAddress, but IPv4 and IPv6 could be used in any case (this is temporary meanwhile the Issue mentioned down bellow is open).

**Which issue(s) this PR fixes** :
Fixes #[OCPBUGS-15331](https://issues.redhat.com/browse/OCPBUGS-15331)
Fixes #[OCPBUGS-16049](https://issues.redhat.com/browse/OCPBUGS-16049)

## Caveats:

- https://github.com/kubernetes/enhancements/issues/2438 - This issue prevents the KAS to support dual stack and the `AdvertiseAddress` even being an slice, it only supports only one element (if more provided it will take the last one as the AdvertiseAddress).

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

File: hypershift-operator/controllers/nodepool/haproxy.go [hunk 1]
```diff
@@ -97,7 +99,20 @@ func (r *NodePoolReconciler) reconcileHAProxyIgnitionConfig(ctx context.Context,
...
 	}
 
-	apiServerInternalAddress := config.DefaultAdvertiseAddress
+	// This give support for HTTP Proxy on IPv6 scenarios
+	defaultAddresses := &util.DefaultAdvIps{
+		IPv4: config.DefaultAdvertiseIPv4Address,
+		IPv6: config.DefaultAdvertiseIPv6Address,
+	}
+
+	firstSvcAddresses := util.GetIpsForSvcNetsWithDefaults(hcluster.Spec.Networking.ServiceNetwork, defaultAddresses)
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/haproxy.go:108]: why is this using the SVC address as the apiServerInternalAddress?

File: support/util/util.go [hunk 2]
```diff
@@ -245,3 +246,69 @@ func ConvertImageRegistryOverrideStringToMap(envVar string) map[string][]string
...
+
+	// I cannot think in a scenario where the serviceNetworks are not rightly parseable
+	// but as a secure measure, if there is not any stored addresses at this point, we set
+	// IPv4 and IPv6 as a default ones.
+	if len(addresses) == 0 {
+		fmt.Println("error parsing defined ServiceNetworkEntries, defaulting AdvertiseAddress to IPv4:", defaults.IPv4)
+		addresses = append(addresses, defaults.IPv4)
+	}
+
+	finalList := RemoveDuplicatesFromList(addresses)
```
Comment (muraee) [inline support/util/util.go:291]: your list can only have 1 item at most, why do you need  `RemoveDuplicatesFromList(addresses)` here?

File: support/util/networking.go [hunk 3]
```diff
@@ -121,3 +130,43 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+	}
+
+	// If there is not any defined AdvertiseAddress we should default them.
+	if len(addresses) == 0 {
+		fmt.Println("AdvertiseAddress or AdvertiseAddresses not found in hcp, using default values")
+		addresses = GetIpsForSvcNetsWithDefaults(hcp.Spec.Networking.ServiceNetwork, defaults)
+	}
+
+	// Clean duplicates from the slice
+	finalList := RemoveDuplicatesFromList(addresses)
```
Comment (muraee) [inline support/util/networking.go:155]: why not just use a map here for `addresses`, and then return the keys.

File: api/v1beta1/hostedcluster_types.go [hunk 4]
```diff
@@ -580,12 +580,20 @@ type CIDRBlock string
...
 	// server. This is an address associated with the loopback adapter of each
-	// node. If not specified, 172.20.0.1 is used.
+	// node. If not specified, 172.20.0.1 or fd02::1 are used.
 	AdvertiseAddress *string `json:"advertiseAddress,omitempty"`
 
+	// AdvertiseAddresses are the addresses that nodes will use to talk to the API
+	// server. These are one or more addresses associated with the loopback adapter of each
+	// node. If not specified, 172.20.0.1 or/and fd02::1 are used.
+	// This variable differs from the above one because this will be used in dual stak scenarios.
+	// NOTE: Both of them could be used at the same time and the controller will take care of them
```
Comment (muraee) [inline api/v1beta1/hostedcluster_types.go:594]: > // NOTE: Both of them could be used at the same time and the controller will take care of them
is this note still true with the CEL validation in place?

File: support/util/networking.go (hunk 3 above)
Comment (jparrill) [reply support/util/networking.go:155]: Just to avoid human errors including repeated CIDRs in the spec

File: api/v1beta1/hostedcluster_types.go (hunk 4 above)
Comment (jparrill) [reply api/v1beta1/hostedcluster_types.go:594]: Oh you're right, let me fix this

File: hypershift-operator/controllers/nodepool/haproxy.go (hunk 1 above)
Comment (jparrill) [reply hypershift-operator/controllers/nodepool/haproxy.go:108]: If I'm not wrong `config.DefaultAdvertiseAddress` is the first usable address of the first found service network. In the [current implementation](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/haproxy.go#L100) we are setting that as the `apiServerInternalAddress`, so I've modified the code to follow that path, having in mind the changes done.

File: support/util/networking.go (hunk 3 above)
Comment (jparrill) [reply support/util/networking.go]: I ended commenting that line, I have that prepared to handle the situation when the issue gets solved.

File: api/v1beta1/hostedcluster_types.go [hunk 5]
```diff
@@ -610,7 +610,7 @@ type CIDRBlock string
 
 // APIServerNetworking specifies how the APIServer is exposed inside a cluster
 // node.
-// +kubebuilder:validation:XValidation:rule="(has(self.advertiseAddress) && self.advertiseAddresses.size() == 0) || (!has(self.advertiseAddress) && self.advertiseAddresses.size() > 0)", message="AdvertiseAddress and AdvertiseAddresses are exclusive between themselves"
+// +kubebuilder:validation:XValidation:rule="(!has(self.advertiseAddress) && !has(self.advertiseAddresses)) || (has(self.advertiseAddress) && !has(self.advertiseAddresses)) || (!has(self.advertiseAddress) && (has(self.advertiseAddresses) && self.advertiseAddresses.size() > 0))", message="AdvertiseAddress and AdvertiseAddresses are exclusive between themselves"
```
Comment (enxebre) [inline api/v1beta1/hostedcluster_types.go:613]: This doesn't belong to this commit. Let's squash both commits if that's easier

File: support/util/util.go [hunk 6]
```diff
@@ -260,3 +260,23 @@ func IsIPv4(cidr string) (bool, error) {
 		return false, nil
 	}
 }
+
+// GetFirstUsableIP returns the first usable IP in both, IPv4 and IPv6 stacks.
+func GetFirstUsableIP(cidr string) (string, error) {
+	_, ipNet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("error validating the incoming CIDR %s: %v", cidr, err)
```
Comment (enxebre) [inline support/util/util.go:268]: please use %w so the error is wrapped https://pkg.go.dev/golang.org/x/xerrors

File: support/util/networking.go [hunk 7]
```diff
@@ -121,3 +130,56 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+
+	// Clean duplicates from the slice
+	// todo (jparrill): Uncomment this and return the final list,
+	// once the issue https://github.com/kubernetes/enhancements/issues/2438 gets solved
+	//finalList := RemoveDuplicatesFromList(addresses)
+
+	return addresses
+}
+
+func RemoveDuplicatesFromList(addresses []string) []string {
```
Comment (enxebre) [inline support/util/networking.go:173]: let's don't commit code we don't consume yet

File: support/util/networking.go [hunk 8]
```diff
@@ -121,3 +130,56 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+	}
+
+	// If there is not any defined AdvertiseAddress we should default them.
+	if len(addresses) == 0 {
+		fmt.Println("AdvertiseAddress or AdvertiseAddresses not found in hcp, using default values")
+		ipv4, err := IsIPv4(hcp.Spec.Networking.ServiceNetwork[0].CIDR.String())
+		if err != nil {
+			// if there is an error validating the ServiceNetworkCIDR we use the IPv4 default ip
+			fmt.Printf("error checking the ServiceNetworkCIDRs: %v", err)
+			return []string{defaults.IPv4}
```
Comment (enxebre) [inline support/util/networking.go:155]: why is this ignoring error and returning default?

File: support/util/networking.go [hunk 9]
```diff
@@ -121,3 +130,56 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+		addresses = append(addresses, *hcpAdvAddress)
+	}
+
+	// If there is not any defined AdvertiseAddress we should default them.
+	if len(addresses) == 0 {
+		fmt.Println("AdvertiseAddress or AdvertiseAddresses not found in hcp, using default values")
+		ipv4, err := IsIPv4(hcp.Spec.Networking.ServiceNetwork[0].CIDR.String())
+		if err != nil {
+			// if there is an error validating the ServiceNetworkCIDR we use the IPv4 default ip
+			fmt.Printf("error checking the ServiceNetworkCIDRs: %v", err)
```
Comment (enxebre) [inline support/util/networking.go:154]: please remove debug print

File: support/util/networking.go [hunk 10]
```diff
@@ -121,3 +130,56 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+	}
+
+	if hcpAdvAddress := AdvertiseAddress(hcp); hcpAdvAddress != nil {
+		fmt.Println("AdvertiseAddress found:", hcpAdvAddress)
+		addresses = append(addresses, *hcpAdvAddress)
+	}
+
+	// If there is not any defined AdvertiseAddress we should default them.
+	if len(addresses) == 0 {
+		fmt.Println("AdvertiseAddress or AdvertiseAddresses not found in hcp, using default values")
```
Comment (enxebre) [inline support/util/networking.go:150]: please remove debug print

File: support/util/networking.go [hunk 11]
```diff
@@ -121,3 +130,56 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
 	}
 	return nil
 }
+
+func SetAdvertiseAddresses(hcp *hyperv1.HostedControlPlane, defaults *DefaultAdvIps) []string {
+	addresses := make([]string, 0)
+
+	// User defined AdvertiseAddress/AdvertiseAddresses
+	if hcpAdvAddresses := AdvertiseAddresses(hcp); hcpAdvAddresses != nil {
```
Comment (enxebre) [inline support/util/networking.go:138]: Why is this if needed? why don't we just `addresses = AdvertiseAddresses(hcp)`?
then only if addresses is still empty you run AdvertiseAddress(hcp), then only if adresses stil empty you return default.

File: control-plane-operator/controllers/hostedcontrolplane/pki/params.go [hunk 12]
```diff
@@ -65,6 +65,22 @@ func NewPKIParams(hcp *hyperv1.HostedControlPlane,
...
-	p.NodeInternalAPIServerIP = util.AdvertiseAddressWithDefault(hcp, config.DefaultAdvertiseAddress)
+
+	// If the first serviceCIDR is an IPv4 we need to set the config.DefaultAdvertiseIPv4Address
+	// as fake IP in the node to access the haproxy exposed as kube-api-server-proxy
+	// Even with that, we cannot set more than one AdvertiseAddress so both
+	// are not supported at the same time.
+	// Check this for more info: https://github.com/kubernetes/enhancements/issues/2438
+	ipv4, err := util.IsIPv4(p.ServiceCIDR)
+	if err != nil {
+		fmt.Printf("error checking the ServiceNetworkCIDRs: %v", err)
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/pki/params.go:89]: why is this error being ignored? if it's intentional please use the logger to print

File: control-plane-operator/controllers/hostedcontrolplane/kas/params.go [hunk 13]
```diff
@@ -388,7 +395,7 @@ type KubeAPIServerConfigParams struct {
 	InternalRegistryHostName     string
 	ExternalRegistryHostNames    []string
 	DefaultNodeSelector          string
-	AdvertiseAddress             string
+	AdvertiseAddress             []string
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/kas/params.go:398]: AdvertiseAddresses?

File: support/util/util.go [hunk 14]
```diff
@@ -260,3 +260,23 @@ func IsIPv4(cidr string) (bool, error) {
 		return false, nil
 	}
 }
+
+// GetFirstUsableIP returns the first usable IP in both, IPv4 and IPv6 stacks.
```
[the remark this answers is not in the document]
Comment (jparrill) [reply support/util/util.go:264]: https://github.com/openshift/hypershift/pull/2779/files#diff-d459531b830f5829c47c3d5b135a2a8c8458c80c7d58c8750aa20403d28c3898R31

File: support/util/networking.go (hunk 8 above)
Comment (jparrill) [reply support/util/networking.go:149]: This is the last mile, the `ServiceNetworkCIDRs` should be validated at this point, if not means that the api machinery didn't validate properly the networks and that should be happened before. This is only to default the value in case of any other validations failed during the HC deployment.

File: control-plane-operator/controllers/hostedcontrolplane/pki/params.go (hunk 12 above)
Comment (jparrill) [reply control-plane-operator/controllers/hostedcontrolplane/pki/params.go:89]: Here 3 scenarios could happen:

- IsIPv4 - This is an IPv4 and we set the proper default address
- !IsIPv4 - This is an IPv6 and we set the proper default address
- Error on validation - This could happen because the ServiceNetwork is not validated, which is something that apimachinery will not allow. This should not happen or at least I don't see how this scenario could happen. In this case it will default the IPv6 address because booleans are initialized as false by default.

File: control-plane-operator/controllers/hostedcontrolplane/kas/params.go (hunk 13 above)
Comment (jparrill) [reply control-plane-operator/controllers/hostedcontrolplane/kas/params.go:398]: Nope, I've mixed both AdvertiseAddresses and AdvertiseAddress in the same param. The behaviour should remain the same because for now we are setting only 1 ip address. We can change the logic to support multi address in this parameter when KAS has support for that (https://github.com/kubernetes/enhancements/issues/2438)

File: support/util/networking.go (hunk 11 above)
Comment (jparrill) [reply support/util/networking.go]: Makes sense. I've modified the function in that way and also I've added some UnitTests over it because I think they was necessary.

File: support/util/networking.go [hunk 15]
```diff
@@ -121,3 +130,56 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+			addresses = append(addresses, defaults.IPv4)
+		} else {
+			addresses = append(addresses, defaults.IPv6)
+		}
+	}
+
+	// Clean duplicates from the slice
+	// todo (jparrill): Uncomment this and return the final list,
+	// once the issue https://github.com/kubernetes/enhancements/issues/2438 gets solved
+	//finalList := RemoveDuplicatesFromList(addresses)
```
[the remark this answers is not in the document]
Comment (jparrill) [reply support/util/networking.go]: Yep, if the user adds a duplicated entry here, I think we should validate that in the networking validation side. I don't think is should belong to this PR but a follow up one. WDYT?

File: control-plane-operator/controllers/hostedcontrolplane/kas/params.go (hunk 13 above)
Comment (muraee) [reply control-plane-operator/controllers/hostedcontrolplane/kas/params.go:398]: even though it holds only 1 ip, I think it should be renamed to AdvertiseAddresses since we changed the type to a slice.

Comment (jparrill) [reply control-plane-operator/controllers/hostedcontrolplane/kas/params.go]: Ok makes sense. I've modified the code and tests accordingly

File: api/v1beta1/hostedcluster_types.go [hunk 16]
```diff
@@ -610,12 +610,22 @@ type CIDRBlock string
...
 // APIServerNetworking specifies how the APIServer is exposed inside a cluster
 // node.
+// +kubebuilder:validation:XValidation:rule="(!has(self.advertiseAddress) && !has(self.advertiseAddresses)) || (has(self.advertiseAddress) && !has(self.advertiseAddresses)) || (!has(self.advertiseAddress) && (has(self.advertiseAddresses) && self.advertiseAddresses.size() > 0))", message="AdvertiseAddress and AdvertiseAddresses are exclusive between themselves"
 type APIServerNetworking struct {
 	// AdvertiseAddress is the address that nodes will use to talk to the API
 	// server. This is an address associated with the loopback adapter of each
-	// node. If not specified, 172.20.0.1 is used.
+	// node. If not specified, the controller will look for the first IP on each
+	// CIDR (IPv4 and IPv6). If the controller cannot infer the first IP of the CIDR,
+	// the default values will be set 172.20.0.1 or/and fd00::1.
```
Comment (csrwng) [inline api/v1beta1/hostedcluster_types.go:619]: Please update this comment to say that if we don't specify the IP, then the default value will be used (not the first of the CIDR)

File: api/v1beta1/hostedcluster_types.go [hunk 17]
```diff
@@ -610,12 +610,22 @@ type CIDRBlock string
...
 	AdvertiseAddress *string `json:"advertiseAddress,omitempty"`
 
+	// AdvertiseAddresses are the addresses that nodes will use to talk to the API
+	// server. These are one or more addresses associated with the loopback adapter of each
+	// node. If not specified, the controller will look for the first IP on each CIDR (IPv4 and IPv6).
+	// If the controller cannot infer the first IP of the CIDR, the default values will be set 172.20.0.1 or/and fd00::1.
+	// This variable differs from the above one because this will be used in dual stack scenarios.
+	AdvertiseAddresses []string `json:"advertiseAddresses,omitempty"`
```
Comment (csrwng) [inline api/v1beta1/hostedcluster_types.go:627]: Unless we can use the multiple advertise addresses, we should not introduce a new field here. Until now, the kube-apiserver only accepts a single advertise IP.

**Review by csrwng — COMMENTED** [review]
Some comments, thx!

File: api/v1beta1/hostedcluster_types.go (hunk 17 above)
Comment (csrwng) [reply api/v1beta1/hostedcluster_types.go:627]: You should add a comment to the existing field with a link to the Kubernetes issue saying that if that issue gets implemented/resolved, then we will need to modify that field.

Comment (jparrill) [reply api/v1beta1/hostedcluster_types.go]: Makes sense, done

File: support/util/networking.go [hunk 18]
```diff
@@ -121,3 +121,29 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
 	}
 	return nil
 }
+
+func SetAdvertiseAddress(hcp *hyperv1.HostedControlPlane, defaults *DefaultAdvIps) string {
```
Comment (csrwng) [inline support/util/networking.go:125]: This function returns an advertise address, so it seems a more appropriate name would be `GetAdvertiseAddress`

File: support/util/networking.go [hunk 19]
```diff
@@ -121,3 +121,29 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+	var address string
+
+	// User defined AdvertiseAddress
+	if address := AdvertiseAddress(hcp); address != nil {
+		return *address
+	}
+
+	// If there is not any defined AdvertiseAddress we should default them.
+	if len(address) <= 1 {
```
Comment (csrwng) [inline support/util/networking.go:134]: It seems that this condition is unnecessary since address is not set

File: support/util/networking.go [hunk 20]
```diff
@@ -121,3 +121,29 @@ func AllowedCIDRBlocks(hcp *hyperv1.HostedControlPlane) []hyperv1.CIDRBlock {
...
+		ipv4, err := IsIPv4(hcp.Spec.Networking.ServiceNetwork[0].CIDR.String())
+		if err != nil {
```
Comment (csrwng) [inline support/util/networking.go:136]: nit: can be combined with the condition below
if err != nil || ipv4 { ...

File: control-plane-operator/controllers/hostedcontrolplane/pki/kas.go [hunk 21]
```diff
@@ -40,10 +45,14 @@ func ReconcileKASServerCertSecret(secret, ca *corev1.Secret, ownerRef config.Own
 		fmt.Sprintf("%s.%s.svc", svc.Name, svc.Namespace),
 		fmt.Sprintf("%s.%s.svc.cluster.local", svc.Name, svc.Namespace),
 	}
+	dnsNames = append(dnsNames, svcAddresses...)
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/pki/kas.go:48]: if svcAddresses is a slice of IP's, why add it to dns names?

File: control-plane-operator/controllers/hostedcontrolplane/kas/params.go [hunk 22]
```diff
@@ -103,7 +103,14 @@ func NewKubeAPIServerParams(ctx context.Context, hcp *hyperv1.HostedControlPlane
 		params.Image = hcp.Spec.Configuration.Image
 		params.Scheduler = hcp.Spec.Configuration.Scheduler
 	}
-	params.AdvertiseAddress = util.AdvertiseAddressWithDefault(hcp, config.DefaultAdvertiseAddress)
+
+	defaultAddresses := &util.DefaultAdvIps{
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/kas/params.go:107]: It seems overkill to have a data structure for defaults and to pass them to the SetAdvertiseAddress function.
They can simply be separate function parameters.

File: control-plane-operator/controllers/hostedcontrolplane/kas/params_test.go [hunk 23]
```diff
@@ -48,21 +53,27 @@ func TestNewAPIServerParamsAPIAdvertiseAddressAndPort(t *testing.T) {
...
 			hcp.Spec.Services = []hyperv1.ServicePublishingStrategyMapping{test.apiServiceMapping}
-			hcp.Spec.Networking.APIServer = &hyperv1.APIServerNetworking{Port: test.port, AdvertiseAddress: test.advertiseAddress}
+			ipnet, err := ipnet.ParseCIDR(test.serviceNetworkCIDR)
+			g.Expect(err).NotTo(HaveOccurred(), "error parsing ServiceNetworkCIDR")
+			hcp.Spec.Networking.ServiceNetwork = append(hcp.Spec.Networking.ServiceNetwork, hyperv1.ServiceNetworkEntry{CIDR: *ipnet})
+			hcp.Spec.Networking.APIServer = &hyperv1.APIServerNetworking{Port: test.port, AdvertiseAddress: pointer.String(test.advertiseAddress)}
 			p := NewKubeAPIServerParams(context.Background(), hcp, imageProvider, "", 0, "", 0, false)
-			g := NewGomegaWithT(t)
-			g.Expect(p.AdvertiseAddress).To(Equal(test.expectedAddress))
+			if len(p.AdvertiseAddress) > 0 {
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/kas/params_test.go:74]: I would think it makes sense to check `if len(test.expectedAddress) > 0` here

File: control-plane-operator/controllers/hostedcontrolplane/pki/params.go (hunk 12 above)
Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/pki/params.go:89]: We should be using the logger if possible to print out the error message... not fmt.Printf

File: hypershift-operator/controllers/nodepool/haproxy.go [hunk 24]
```diff
@@ -97,7 +99,19 @@ func (r *NodePoolReconciler) reconcileHAProxyIgnitionConfig(ctx context.Context,
 		return "", true, fmt.Errorf("release image doesn't have a %s image", haProxyRouterImageName)
 	}
 
-	apiServerInternalAddress := config.DefaultAdvertiseAddress
+	// This give support for HTTP Proxy on IPv6 scenarios
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool/haproxy.go:102]: s/give/provides/

File: support/util/networking_test.go [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+func TestSetAdvertiseAddress(t *testing.T) {
+	type args struct {
+		hcp      *hyperv1.HostedControlPlane
+		defaults *DefaultAdvIps
+	}
+	tests := []struct {
+		name string
+		args args
```
Comment (csrwng) [inline support/util/networking_test.go:19]: why not include hcp directly here? 
also why do you need to specify defaults for every case?

File: support/util/types.go [hunk 26]
```diff
@@ -0,0 +1,6 @@
+package util
+
+type DefaultAdvIps struct {
```
Comment (csrwng) [inline support/util/types.go:3]: As mentioned above, this struct seems unnecessary

File: support/util/networking.go (hunk 18 above)
Comment (csrwng) [reply support/util/networking.go:125]: In general, this function is distinguishing between ipv4 and ipv6 and then essentially doing what `AdvertiseAddressWithDefault` does, so maybe you should call that function from here.

File: support/util/util.go [hunk 27]
```diff
@@ -260,3 +260,23 @@ func IsIPv4(cidr string) (bool, error) {
...
+	_, ipNet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("error validating the incoming CIDR %s: %w", cidr, err)
+	}
+
+	// If we receive a CIDR which is not a network address
+	ip := ipNet.IP
+	if ip.To4() != nil {
+		// IPv4
+		ip[net.IPv4len-1]++
```
Comment (csrwng) [inline support/util/util.go:275]: both the ipv4 and ipv6 cases seem to be doing the same thing, so ip[len(ip)-1]++ should work for both, I don't think there's a need to differentiate.

File: support/util/util.go [hunk 28]
```diff
@@ -260,3 +260,23 @@ func IsIPv4(cidr string) (bool, error) {
 		return false, nil
 	}
 }
+
+// GetFirstUsableIP returns the first usable IP in both, IPv4 and IPv6 stacks.
+func GetFirstUsableIP(cidr string) (string, error) {
```
Comment (csrwng) [inline support/util/util.go:265]: nit: s/GetFirstUsableIP/FirstUsableIP/

**Review by csrwng — COMMENTED** [review]
Thank you Juan, some additional comments

File: control-plane-operator/controllers/hostedcontrolplane/pki/kas.go (hunk 21 above)
Comment (jparrill) [reply control-plane-operator/controllers/hostedcontrolplane/pki/kas.go:48]: Because this way is how was set in the Mario's env.

OCP Original:
```
DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, DNS:kubernetes.default.svc.cluster.local, DNS:openshift, DNS:openshift.default, DNS:openshift.default.svc, DNS:openshift.default.svc.cluster.local, DNS:172.30.0.1, DNS:fd02::1, IP Address:172.30.0.1, IP Address:FD02:0:0:0:0:0:0:1
```

Env:
```
DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, DNS:kubernetes.default.svc.cluster.local, DNS:openshift, DNS:openshift.default, DNS:openshift.default.svc, DNS:openshift.default.svc.cluster.local, DNS:172.30.0.1, DNS:fd02::1, IP Address:172.30.0.1, IP Address:FD02:0:0:0:0:0:0:1
```

New:
```
DNS:localhost, DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, DNS:kubernetes.default.svc.cluster.local, DNS:kube-apiserver, DNS:kube-apiserver.clusters-hosted.svc, DNS:kube-apiserver.clusters-hosted.svc.cluster.local, DNS:api.hosted.dual.lab, DNS:api.hosted.hypershift.local, DNS:172.31.0.1, DNS:fd02::1, IP Address:127.0.0.1, IP Address:0:0:0:0:0:0:0:1, IP Address:172.31.0.1, IP Address:FD02:0:0:0:0:0:0:1
```

File: control-plane-operator/controllers/hostedcontrolplane/pki/params.go (hunk 12 above)
Comment (jparrill) [reply control-plane-operator/controllers/hostedcontrolplane/pki/params.go:89]: I've decided to just avoid the error message, because the api machinery will hit the issue before and it will report the error early.

File: support/util/util.go (hunk 27 above)
Comment (jparrill) [reply support/util/util.go:275]: If I do that this is the result in the unitTests:

```
FirstUsableIP() = 2000::, want 2000::1
```

File: support/util/networking_test.go (hunk 25 above)
Comment (jparrill) [reply support/util/networking_test.go:23]: Not following the first comment, fixed the second one.

File: support/util/util.go (hunk 27 above)
Comment (csrwng) [reply support/util/util.go:275]: I tested ip[len(ip)-1]++ for all cases and it passes for me
